### PR TITLE
Added Monitoring Capabilities (Observability-as-code)

### PR DESCRIPTION
**Pull request by bjcmit**
## Summary of the Pull Request
Added the following capabilities that can be configured to be enabled/disabled:
- Azure Dashboard, Alerts, and Diagnostic Setting for Synapse, Cosmos DB, IoT Hub, Event Hub, and Stream Analytics
- Log Analytics Workspace
- Updated Portal Experience
- Updated Doc

## References
No references

## PR Checklist
* [X] CLA signed.

## Validation Steps Performed
Deployed manually

### CODE REVIEW

**Comment by sudivate** [comment]
Please add similar observability elements for other services like Cosmos, SQL, IOT hub etc. included in this data product

File: docs/reference/portal.dataProduct.json [hunk 1]
```diff
@@ -956,9 +1011,11 @@
                 "enableSqlServer": "[steps('generalSettings').generalSettings.enableSqlServer]",
                 "enableCosmos": "[steps('generalSettings').generalSettings.enableCosmos]",
                 "enableStreamAnalytics": "[steps('generalSettings').streamAnalyticsDeploymentSettings.enableStreamAnalytics]",
+                "enableMonitoring": "[steps('generalSettings').monitoringSettings.enableMonitoring]",
+                "dataProductTeamEmail": "[steps('generalSettings').monitoringSettings.dataProductTeamEmail]",
```
Comment (marvinbuss) [inline docs/reference/portal.dataProduct.json:1015]: I believe this needs to be corrected.

File: infra/main.bicep [hunk 2]
```diff
@@ -87,6 +91,11 @@ var iothub001Name = '${name}-iothub001'
 var eventhubNamespace001Name = '${name}-eventhub001'
 var streamanalytics001Name = '${name}-streamanalytics001'
 var streamanalyticscluster001Name = '${name}-streamanalyticscluster001'
+var loganalyticsName = '${name}-loganalytics'
+var dataEmailActionGroup = '${name}-emailactiongroup'
+var synapsePipelineFailedAlertName = '${synapse001Name}-failedalert'
+var synapseScope = '${subscription().id}/resourceGroups/${resourceGroup().name}/providers/Microsoft.Synapse/workspaces/${synapse001Name}'
```
Comment (marvinbuss) [inline infra/main.bicep:97]: This is not required. Use output of deployment.

File: infra/main.bicep [hunk 3]
```diff
@@ -87,6 +91,11 @@ var iothub001Name = '${name}-iothub001'
 var eventhubNamespace001Name = '${name}-eventhub001'
 var streamanalytics001Name = '${name}-streamanalytics001'
 var streamanalyticscluster001Name = '${name}-streamanalyticscluster001'
+var loganalyticsName = '${name}-loganalytics'
```
Comment (marvinbuss) [inline infra/main.bicep:94]:
```suggestion
var logAnalytics001Name = '${name}-loganalytics001'
```

File: infra/main.bicep [hunk 4]
```diff
@@ -101,6 +110,52 @@ module keyVault001 'modules/services/keyvault.bicep' = {
   }
 }
 
+module logAnalytics001 'modules/services/loganalytics.bicep' = if(enableMonitoring) {
+  name: 'logAnalytics001'
+  scope: resourceGroup()
+  params: {
+    location: location
+    tags: tagsJoined
+    loganalyticsName: loganalyticsName
```
Comment (marvinbuss) [inline infra/main.bicep:119]:
```suggestion
    logAnalyticsName: loganAlytics001Name  
```

File: infra/modules/services/alerts.bicep [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+// Parameters
+param location string
+param tags object
+param synapsePipelineFailedAlertName string
+param synapseScope string
+param dataEmailActionGroup string
+param dataProductTeamEmail string
+
+
```
Comment (marvinbuss) [inline infra/modules/services/alerts.bicep:15]:
```suggestion

```

File: infra/modules/services/diagnosticsettings.bicep [hunk 6]
```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+// This template is used to setup diagnostic settings.
+targetScope = 'resourceGroup'
+
+// Parameters
+param loganalyticsName string
```
Comment (marvinbuss) [inline infra/modules/services/diagnosticsettings.bicep:8]:
```suggestion
param logAnalyticsName string
```

File: infra/modules/services/diagnosticsettings.bicep [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+resource synapsebigdatapool 'Microsoft.Synapse/workspaces/bigDataPools@2021-06-01' existing = {
+  parent: synapseworkspace
+  name: synapseSparkPoolName
+}
+
+resource logAnalyticsWorkspace 'Microsoft.OperationalInsights/workspaces@2020-08-01' existing = {
+  name: loganalyticsName
+}
+
+resource diagnosticSetting1 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = {
```
Comment (marvinbuss) [inline infra/modules/services/diagnosticsettings.bicep:32]: Let's be consistent here. 👍 
```suggestion
resource diagnosticSetting001 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = {
```

**Review by marvinbuss — CHANGES_REQUESTED** [review]
Please see below. Please also check in ARM after the bicep build.

**Comment by sudivate** [comment]
Please fix formatting on all changed files

**Comment by sudivate** [comment]
Add [this](https://github.com/Azure/data-product-batch/blob/f639ace2ca46e3b85e3fd381b3cdde4b5ce1b5ea/.github/linters/.arm-ttk.psd1#L14) to ARM test toolkit to pass the linter in the PR

**Comment by bjcmit** [comment]
> Please fix formatting on all changed file

I looked through some of the files and I don't see any formatting issues. Could you please provide any example?

**Comment by sudivate** [comment]
> > Please fix formatting on all changed file
> 
> I looked through some of the files and I don't see any formatting issues. Could you please provide any example?

If you are using VS code, use SHIFT + Alt + F to see the changes

File: docs/reference/portal.dataProduct.json [hunk 8]
```diff
@@ -560,7 +560,7 @@
                         },
                         {
                             "name": "generalSettings",
-                            "label": "General Settings",
+                            "label": "Analytics Service Settings",
```
Comment (sudivate) [inline docs/reference/portal.dataProduct.json:563]: [Portal Template Invalid](/#blade/Microsoft_Azure_CreateUIDef/CustomDeploymentBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fdata-product-streaming%2Fbjcmit%2Fdata_observability%2Finfra%2Fmain.json/uiFormDefinitionUri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fdata-product-streaming%2Fbjcmit%2Fdata_observability%2Fdocs%2Freference%2Fportal.dataProduct.json)

File: infra/modules/services/diagnosticsettings.bicep [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+      {
+        category: 'BigDataPoolAppsEnded'
+        enabled: true
+      }
+    ]
+  }
+}]
+
+resource diagnosticSetting4 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = {
+  scope: cosmosdb
```
Comment (sudivate) [inline infra/modules/services/diagnosticsettings.bicep:145]: Will this work if the  cosmos id set to false in params ?

File: infra/modules/services/diagnosticsettings.bicep [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+      {
+        category: 'ControlPlaneRequests'
+        enabled: true
+      }
+    ]
+  }
+}
+
+resource diagnosticSetting5 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = {
+  scope: iothub
```
Comment (sudivate) [inline infra/modules/services/diagnosticsettings.bicep:175]: Will this work if to set to false in params ?

**Review by sudivate — CHANGES_REQUESTED** [review]

File: docs/reference/portal.dataProduct.json (hunk 8 above)
Comment (bjcmit) [reply docs/reference/portal.dataProduct.json:563]: What is your suggested change?

Comment (bjcmit) [reply docs/reference/portal.dataProduct.json:563]: Should the label be "General Settings"?

File: infra/modules/services/diagnosticsettings.bicep (hunk 9 above)
Comment (bjcmit) [reply infra/modules/services/diagnosticsettings.bicep:145]: I get: 
{
  'code': 'InvalidTemplate',
  'message': "Deployment template validation failed: 'The template resource 'Microsoft.Resources/deployments/dashboard' reference      to 'Microsoft.Resources/deployments/cosmos001' requires an API version. Please see https://aka.ms/arm-template for usage details.'.",
  'additionalInfo': [
    {
      'type': 'TemplateViolation',
      'info': {
        'lineNumber': 0,
        'linePosition': 0,
        'path': ''
      }
    }
  ]
}

File: infra/modules/services/diagnosticsettings.bicep (hunk 10 above)
Comment (bjcmit) [reply infra/modules/services/diagnosticsettings.bicep:175]: I get a similar error as when I set enableCosmos to false.

**Review by sudivate — APPROVED** [review]

**Comment by bjcmit** [comment]
> Please add similar observability elements for other services like Cosmos, SQL, IOT hub etc. included in this data product

Added similar observability elements for other services like Cosmos, SQL, IOT hub, etc.

**Comment by marvinbuss** [comment]
[Preview Link](https://portal.azure.com/#blade/Microsoft_Azure_CreateUIDef/CustomDeploymentBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fdata-product-streaming%2Fbjcmit%2Fdata_observability%2Finfra%2Fmain.json/uiFormDefinitionUri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fdata-product-streaming%2Fbjcmit%2Fdata_observability%2Fdocs%2Freference%2Fportal.dataProduct.json)

**Comment by marvinbuss** [comment]
Just tested this in Portal:
image
Have you deployed this successfully?

File: docs/reference/portal.dataProduct.json [hunk 11]
```diff
@@ -612,14 +612,69 @@
                                 {
                                     "name": "enableRoleAssignments",
                                     "label": "Enable role assignments",
-                                    "type": "Microsoft.Common.CheckBox",
+                                    "type": "Microsoft.Common.OptionsGroup",
                                     "visible": true,
```
Comment (marvinbuss) [inline docs/reference/portal.dataProduct.json:616]:
```suggestion
                                    "visible": true,
                                    "defaultValue": "Yes (recommended)",
```

File: docs/reference/portal.dataProduct.json [hunk 12]
```diff
@@ -612,14 +612,69 @@
...
+                                    "type": "Microsoft.Common.OptionsGroup",
+                                    "visible": true,
+                                    "toolTip": "Deploy monitoring components like Log Analytics, Azure dashboards, alerts, and diagnostic settings.",
+                                    "defaultValue": "No",
+                                    "constraints": {
+                                        "allowedValues": [
+                                            {
+                                                "label": "Yes",
+                                                "value": "yes"
+                                            },
```
Comment (marvinbuss) [inline docs/reference/portal.dataProduct.json:663]:
```suggestion
                                    "defaultValue": "Yes (recommended)",
                                    "constraints": {
                                        "allowedValues": [
                                            {
                                                "label": "Yes (recommended)",
                                                "value": "yes"
                                            },
```

File: docs/reference/portal.dataProduct.json [hunk 13]
```diff
@@ -612,14 +612,69 @@
...
+                                        ]
+                                    }
+                                },
+                                {
+                                    "name": "dataProductTeamEmail",
+                                    "label": "Alert Receivers Email ID",
+                                    "type": "Microsoft.Common.TextBox",
+                                    "visible": "[equals(steps('generalSettings').monitoringSettings.enableMonitoring, 'yes')]",
+                                    "defaultValue": "",
+                                    "toolTip": "Specify Email ID to receive monitoring alerts."
```
Comment (marvinbuss) [inline docs/reference/portal.dataProduct.json:677]: Can we apply some checks here to make sure that this is a valid email?
Is the email required?

File: infra/main.bicep [hunk 14]
```diff
@@ -87,6 +91,14 @@ var iothub001Name = '${name}-iothub001'
...
 var streamanalytics001Name = '${name}-streamanalytics001'
 var streamanalyticscluster001Name = '${name}-streamanalyticscluster001'
+var logAnalytics001Name = '${name}-loganalytics001'
+var dataEmailActionGroup = '${name}-emailactiongroup'
+var synapsePipelineFailedAlertName = '${synapse001Name}-failedalert'
+var iothubFailedAlertName = '${iothub001Name}-failedalert'
+var eventhubnamespaceErrorAlertName = '${eventhubNamespace001Name}-erroralert'
+var streamanalyticsErrorAlertName = '${streamanalytics001Name}-erroralert'
+var dashboardName = '${name}-dashboard'
+var database001Name = 'Database001'
```
Comment (marvinbuss) [inline infra/main.bicep:101]: Can you please check and be consistent with the naming convention here:
https://github.com/Azure/data-product-analytics/pull/123

Comment (marvinbuss) [reply infra/main.bicep:101]: Just want to make sure that it is consistent. :)

File: infra/modules/services/sql.bicep [hunk 15]
```diff
@@ -136,3 +137,4 @@ resource sqlserverPrivateEndpointARecord 'Microsoft.Network/privateEndpoints/pri
 
 // Outputs
 output sqlserverId string = sqlserver.id
+output sqlId string = sqlserverDatabase001.id
```
Comment (marvinbuss) [inline infra/modules/services/sql.bicep:140]:
```suggestion
output sqlserverDatabaseId string = sqlserverDatabase001.id
```

**Review by marvinbuss — CHANGES_REQUESTED** [review]
Please see comments below.

**Comment by bjcmit** [comment]
> Just tested this in Portal: image Have you deployed this successfully?

I have deployed successfully via script, but not via Portal.

File: docs/reference/portal.dataProduct.json (hunk 13 above)
Comment (bjcmit) [reply docs/reference/portal.dataProduct.json:678]: There are checks in the bicep files to make sure the email is valid.

File: infra/main.bicep (hunk 14 above)
Comment (bjcmit) [reply infra/main.bicep:101]: Made sure it is consistent.

**Comment by bjcmit** [comment]
> > Just tested this in Portal: image Have you deployed this successfully?
> 
> I have deployed successfully via script, but not via Portal.

Resolved template validation error

**Review by marvinbuss — APPROVED** [review]
LGTM!

**Pull request merged**